### ISSUE TITLE
Distribute roots.pem with Python package

### ISSUE THREAD

**Issue by dhermes** [grpc/grpc#4224]
Inspired by #3805

I've been using system certs in a hacky way via:
https://gist.github.com/dhermes/bbc5b7be1932bfffae77

An OS X user filed https://github.com/dhermes/gcloud-python-bigtable/issues/21 and I tried to address with another hack:
https://github.com/dhermes/gcloud-python-bigtable/commit/d16fa35c4070445d38a5103914c1c207487eb7ed

However, I'd like to be able to create `ssl_client_credentials` without having to know where the roots are stored. Currently just to make an request to a secure URI I [have to do][1]:

```python
from grpc.beta import implementations
root_certificates = 'PEM contents loaded from some file'
client_credentials = implementations.ssl_client_credentials(
    root_certificates, private_key=None, certificate_chain=None)
```

/cc @nathanielmanistaatgoogle @jgeewax @tseaver 

[1]: https://github.com/dhermes/gcloud-python-bigtable/blob/d16fa35c4070445d38a5103914c1c207487eb7ed/gcloud_bigtable/_helpers.py#L271-L272

**Comment by nathanielmanistaatgoogle** [comment]
If you pass `None` as the `root_certificates`, do you find that the "system default" root certificates are found and used? The documentation for that parameter does read `The PEM-encoded root certificates or None to ask for them to be retrieved from a default location.`, doesn't it?

**Comment by dhermes** [comment]
Fingers crossed. Checking now :grinning:

**Comment by dhermes** [comment]
```python
In [8]: implementations.ssl_client_credentials?
Signature: implementations.ssl_client_credentials(root_certificates, private_key, certificate_chain)
Docstring:
Creates a ClientCredentials for use with an SSL-enabled Channel.

Args:
  root_certificates: The PEM-encoded root certificates or None to ask for
    them to be retrieved from a default location.
  private_key: The PEM-encoded private key to use or None if no private key
    should be used.
  certificate_chain: The PEM-encoded certificate chain to use or None if no
    certificate chain should be used.

Returns:
  A ClientCredentials for use with an SSL-enabled Channel.
File:      /usr/local/lib/python2.7/dist-packages/grpc/beta/implementations.py
Type:      function
```

------------

Can I get a RTFM @dhermes?

------------

Unfortunately there is still an issue

```python
In [1]: from gcloud_bigtable import Client

In [2]: client = Client(project='my-project')

In [3]: client.start()
E1201 14:54:31.379212042   13770 file.c:81]                  Could not open file /usr/share/grpc/roots.pem (error = No such file or directory).
E1201 14:54:31.379231520   13770 security_connector.c:581]   Could not get default pem root certs.
```

------------

I installed the core via `apt-get install libgrpc-dev` on Ubuntu 14.04 and then just `pip install grpcio==0.11.0b1`.

**Comment by nathanielmanistaatgoogle** [comment]
@jboeuf and @ctiller: what's the user experience supposed to be here? How is it that @dhermes should come to have a `/usr/share/grpc/roots.pem` file (or for the core to find an acceptable file elsewhere)?

**Comment by soltanmm** [comment]
@dhermes Is that output spat out and the client goes on to eventually find something and work (i.e. it's just being verbose) or is that actually a failure?

**Comment by dhermes** [comment]
Good call. I'll try to execute something.

**Comment by dhermes** [comment]
```python
In [1]: from gcloud_bigtable import Client

In [2]: client = Client(project='my-project', admin=True)

In [3]: client.start()
E1201 15:04:23.975698434   14455 file.c:81]                  Could not open file /usr/share/grpc/roots.pem (error = No such file or directory).
E1201 15:04:23.975719329   14455 security_connector.c:581]   Could not get default pem root certs.
E1201 15:04:23.976557546   14455 security_connector.c:581]   Could not get default pem root certs.
E1201 15:04:23.977202667   14455 security_connector.c:581]   Could not get default pem root certs.
E1201 15:04:23.977640730   14455 security_connector.c:581]   Could not get default pem root certs.

In [4]: from grpc.framework.interfaces.face.face import NetworkError

In [5]: try:
   ...:     result = client.list_zones()
   ...: except NetworkError as exc:
   ...:     print 'Failed'
   ...:     

Failed

In [6]: exc.code
Out[6]: <StatusCode.INVALID_ARGUMENT: 3>

In [7]: exc.details
Out[7]: 'Failed to create security connector.'
```

**Comment by nathanielmanistaatgoogle** [comment]
@dhermes: it seems like a safe assumption but just to draw it out explicitly: that's a code sequence that works perfectly well if you explicitly provide a non-`None` `root_certificates` on channel creation?

**Comment by dhermes** [comment]
Yup. I [hard code][1] the system paths used on Debian Linux and OS X (with the one on OS X installed with `homebrew`'s version of `openssl`).

[1]: https://github.com/dhermes/gcloud-python-bigtable/blob/d16fa35c4070445d38a5103914c1c207487eb7ed/gcloud_bigtable/_helpers.py#L52-L57

**Comment by jboeuf** [comment]
Could the python packaging script do something similar to what nodejs did here in order to include the roots included in the grpc source tree (in `etc/`)?
See https://github.com/grpc/grpc/pull/3846/files

**Comment by soltanmm** [comment]
@jboeuf I'm confused - I thought we were trying to have the user be able to load from a default location? How does that link to distributing more files?

**Comment by dhermes** [comment]
It seems to be a hard problem. The `certifi` project (https://certifi.io/en/latest/) just ships their own certs file rather than trying to make an OS specific mapping to default locations. There is some [discussion][1] of this on the Python bug tracker.

[1]: https://bugs.python.org/issue13655#msg192601

**Comment by soltanmm** [comment]
@dhermes The last message appears to say that they weren't going to distribute (back in January of this year). They went and distributed later on? Before I try to do some digging, do you already have other relevant discussion links handy to reconcile that last bit and what appears to be the case now?

**Comment by dhermes** [comment]
The Python bug tracker (i.e. `python-dev`) `!=` the open source project `certifi`.

-----

The specific comment I linked to was about managing a list of default system paths to certs (which BTW doesn't help on Windows, or OS X, since the OS X path in the thread no longer is shipped with OS X):

```python
cert_paths = [
    # Debian, Ubuntu, Arch, SuSE
    # NetBSD (security/mozilla-rootcerts)
    "/etc/ssl/certs/",
    # Debian, Ubuntu, Arch: maintained by update-ca-certificates
    "/etc/ssl/certs/ca-certificates.crt",
    # Red Hat 5+, Fedora, Centos
    "/etc/pki/tls/certs/ca-bundle.crt",
    # Red Hat 4
    "/usr/share/ssl/certs/ca-bundle.crt",
    # FreeBSD (security/ca-root-nss package)
    "/usr/local/share/certs/ca-root-nss.crt",
    # FreeBSD (deprecated security/ca-root package, removed 2008)
    "/usr/local/share/certs/ca-root.crt",
    # FreeBSD (optional symlink)
    # OpenBSD
    "/etc/ssl/cert.pem",
    # Mac OS X
    "/System/Library/OpenSSL/certs/cert.pem",
    ]
```

**Comment by dhermes** [comment]
The reason I mentioned `certifi` is because that group of people gave up on managing system paths and just ships a `certs` file in their source code. This is the same thing being suggested here. So either shipping it with `libgrpc` core or ship it with the language specific libraries is the equivalent here. (I'm certainly out of my depth in discussing what is secure or correct, I was just trying to explain the things I've seen.)

**Comment by jboeuf** [comment]
@soltanmm: we have a list of roots in the grpc source tree that is suitable for publicly trusted certs:
https://github.com/grpc/grpc/blob/master/etc/roots.pem

So what I would be inclined to do is to ship this with the python package as opposed to depend on the roots installed (or not) on the different platforms. Once you have it shipped with the python package, you can do python magic to resolve the path and set the right env var. This is what nodejs did and it looks like it's working well for them. @murgatroid99 can you confirm?

**Comment by murgatroid99** [comment]
Yes, nodejs did that. But nodejs also ships with the C core bundled instead of linking against it, so I have no expectation that our users would have the certificates installed in `/usr/share/grpc/`.

**Comment by soltanmm** [comment]
That is another little hitch - we require the core package to already be installed. Seeing as how the core determines the expected paths (or, seemingly, not), shouldn't it then be the core package that distributes the required list?

**Comment by jboeuf** [comment]
In that case yes, the roots should come for the installed core package and everything should be working seamlessly thanks to the magic that @nicolasnoble put in the path computation of the default roots file path. 

What's going on right now if you pass `None` to the roots arg of the SSL creds? It might just work...

**Comment by dhermes** [comment]
@jboeuf See my comments above ([1][1], [2][2])

[1]: https://github.com/grpc/grpc/issues/4224#issuecomment-161123739
[2]: https://github.com/grpc/grpc/issues/4224#issuecomment-161125776

**Comment by jboeuf** [comment]
ah yes sorry...

@nicolasnoble, are you the owner of the core package? Does that include the roots? If so, where are they :)? Thanks!

**Comment by soltanmm** [comment]
Hmm, so now that we're going to be distributing the core with the Python installation, the certs should somehow be distributed with the installation?

There'd be the (minor, easily worked around) hitch that by design Python packages are not able to install to arbitrary locations. This brings a question: is it a problem thus that the users would never see the certs on their computers in any standard location?

**Comment by dhermes** [comment]
FWIW I recently discovered the `httplib2` also [distributes its own certs][1] and allows the user to [override][2] if they so choose.

[1]: https://github.com/jcgregorio/httplib2/blob/cf631a73e2f3f43897b65206127ced82382d35f5/python2/httplib2/cacerts.txt
[2]: https://github.com/jcgregorio/httplib2/blob/cf631a73e2f3f43897b65206127ced82382d35f5/python2/httplib2/__init__.py#L187-L195

**Comment by ctiller** [comment]
3 proposals to move forward in the new core embedded world in order of my
preference:
1. See if there's a Python library that can dig out the system certs for us
(maybe we get lucky)
2. Bundle the certs as data with the extension
3. Bundle the certs as code with the extension

All would need a hook for users to override the defaults, and that hook to
be very clearly documented.

We should probably have a well known HTTPS address for our users to fetch
them also. I could even imagine providing an API to fetch whatever we think
is latest.

**Comment by jboeuf** [comment]
FYI, this is how we got the `etc/roots.pem` in the tree.

1. Get the certs from Mozilla:
https://hg.mozilla.org/mozilla-central/raw-file/tip/security/nss/lib/ckfw/builtins/certdata.txt

2. Convert them to PEM using
https://github.com/agl/extract-nss-root-certs.

I'm going to do a refresh shortly (it's been one year or so).

**Comment by jboeuf** [comment]
@ctiller @nicolasnoble should this be in the upcoming 0.12 release?

**Comment by nathanielmanistaatgoogle** [comment]
@soltanmm is the story on this any different since the merge of #4457?

**Comment by soltanmm** [comment]
Fixed in 6d2ef17.
